src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -72,14 +73,42 @@ public abstract partial class Frame : ConnectionContext, IFrameControl
         protected readonly long _keepAliveMilliseconds;
 
         public Frame(ConnectionContext context)
-            : base(context)
         {
-            _pathBase = context.ServerAddress.PathBase;
+            ConnectionContext = context;
+            SocketInput = context.SocketInput;
+            SocketOutput = context.SocketOutput;
+
+            _pathBase = ServerAddress.PathBase;
 
             FrameControl = this;
             _keepAliveMilliseconds = (long)ServerOptions.Limits.KeepAliveTimeout.TotalMilliseconds;
         }
 
+        public ConnectionContext ConnectionContext { get; }
+        public SocketInput SocketInput { get; set; }
+        public ISocketOutput SocketOutput { get; set; }
+        public Action<IFeatureCollection> PrepareRequest
+        {
+            get
+            {
+                return ConnectionContext.PrepareRequest;
+            }
+            set
+            {
+                ConnectionContext.PrepareRequest = value;
+            }
+        }
+
+        protected IConnectionControl ConnectionControl => ConnectionContext.ConnectionControl;
+        protected IKestrelTrace Log => ConnectionContext.ListenerContext.ServiceContext.Log;
+
+        private DateHeaderValueManager DateHeaderValueManager => ConnectionContext.ListenerContext.ServiceContext.DateHeaderValueManager;
+        private ServerAddress ServerAddress => ConnectionContext.ListenerContext.ServerAddress;
+        private KestrelServerOptions ServerOptions => ConnectionContext.ListenerContext.ServiceContext.ServerOptions;
```

Comment:
We actually use this a lot in frame (e.g. MaxRequestLineSize, MaxHeaderCount). Can we make KestrelServerLimits a private field and set it in the ctor instead of having frame access this via `ConnectionContext.ListenerContext.ServiceContext.ServerOptions.Limits` every time? Even if the properties are inlined, that's a lot of indirection for a fairly hot path.